haystack/nodes/reader/farm.py

```diff
@@ -1306,30 +1306,39 @@ def _deduplicate_predictions(self, predictions: List["QAPred"], documents: List[
-                    # Check if answer offsets are within the overlap
-                    if not self._qa_cand_in_overlap(ans, overlap):
-                        continue
+                self.handle_qa_overlap_duplicates(
+                    overlapping_docs, preds_per_doc, pred, relevant_overlaps, ans_idx, ans
+                )
+        return predictions
 
-                    # Check if predictions from overlapping Document are within the overlap
-                    overlapping_doc_pred = preds_per_doc[overlap["doc_id"]]
-                    cur_doc_overlap = [ol for ol in overlapping_docs[overlap["doc_id"]] if ol["doc_id"] == pred.id][0]
-                    for pot_dupl_ans_idx in reversed(range(len(overlapping_doc_pred.prediction))):
-                        pot_dupl_ans = overlapping_doc_pred.prediction[pot_dupl_ans_idx]
-                        if pot_dupl_ans.answer_type != "span":
-                            continue
-                        if not self._qa_cand_in_overlap(pot_dupl_ans, cur_doc_overlap):
-                            continue
+    def handle_qa_overlap_duplicates(self, overlapping_docs, preds_per_doc, pred, relevant_overlaps, ans_idx, ans):
+        for overlap in relevant_overlaps:
+            # Check if answer offsets are within the overlap
+            if not self._qa_cand_in_overlap(ans, overlap):
+                continue
 
-                        # Check if ans and pot_dupl_ans are duplicates
-                        if self._is_duplicate_answer(ans, overlap, pot_dupl_ans, cur_doc_overlap):
-                            # Discard the duplicate with lower score
-                            if ans.confidence < pot_dupl_ans.confidence:
-                                pred.prediction.pop(ans_idx)
-                            else:
-                                overlapping_doc_pred.prediction.pop(pot_dupl_ans_idx)
+                # Check if predictions from overlapping Document are within the overlap
+            overlapping_doc_pred = preds_per_doc[overlap["doc_id"]]
+            cur_doc_overlap = [ol for ol in overlapping_docs[overlap["doc_id"]] if ol["doc_id"] == pred.id][0]
+            breaked = False
+            for pot_dupl_ans_idx in reversed(range(len(overlapping_doc_pred.prediction))):
+                pot_dupl_ans = overlapping_doc_pred.prediction[pot_dupl_ans_idx]
+                if pot_dupl_ans.answer_type != "span":
+                    continue
+                if not self._qa_cand_in_overlap(pot_dupl_ans, cur_doc_overlap):
+                    continue
 
-        return predictions
+                    # Check if ans and pot_dupl_ans are duplicates
+                if self._is_duplicate_answer(ans, overlap, pot_dupl_ans, cur_doc_overlap):
+                    # Discard the duplicate with lower score
+                    if ans.confidence < pot_dupl_ans.confidence:
+                        pred.prediction.pop(ans_idx)
+                        breaked = True
+                        break
```

Comment:
Instead of `break` we should use `return` here to break out of both for loops. The lines `breaked = False` and `breaked = True` can also be removed then. For context, feel free to have a look a my comment on the original issue here: https://github.com/deepset-ai/haystack/issues/4954#issuecomment-1561314712